chain/beacon/chain.go

```diff
@@ -20,40 +18,55 @@ const (
 	defaultNewBeaconBuffer   = 100
 )
 
-// chainStore is a Store that deals with reconstructing the beacons, sync when
-// needed and arranges the head
+// chainStore implements CallbackStore, Syncer and deals with reconstructing the
+// beacons, and sync when needed. This struct is the gateway logic for beacons to
+// be inserted in the database and for replying to beacon requests.
 type chainStore struct {
-	chain.Store
-	conf          *Config
-	l             log.Logger
-	client        net.ProtocolClient
-	safe          *cryptoSafe
-	ticker        *ticker
-	done          chan bool
-	newPartials   chan partialInfo
-	newBeaconCh   chan *chain.Beacon
-	lastInserted  chan *chain.Beacon
-	requestSync   chan likeBeacon
-	nonSyncBeacon chan *chain.Beacon
+	CallbackStore
+	sync        Syncer
+	conf        *Config
+	l           log.Logger
+	client      net.ProtocolClient
+	crypto      *cryptoStore
+	ticker      *ticker
+	done        chan bool
+	newPartials chan partialInfo
+	// catchupBeacons is used to notify the Handler when a node has aggregated a
+	// beacon.
+	catchedupBeacons chan *chain.Beacon
+	// all beacons finally inserted into the store are sent over this cannel for
+	// the aggregation loop to know
+	beaconStoredAgg chan *chain.Beacon
 }
 
-func newChainStore(l log.Logger, conf *Config, client net.ProtocolClient, safe *cryptoSafe, s chain.Store, ticker *ticker) *chainStore {
+func newChainStore(l log.Logger, conf *Config, client net.ProtocolClient, crypto *cryptoStore, store chain.Store, ticker *ticker) *chainStore {
+	// we make sure the chain is increasing monotically
+	as := newAppendStore(store)
+	// we write some stats about the timing when new beacon is saved
+	ds := newDiscrepancyStore(as, l, crypto.GetGroup())
+	// we can register callbacks on it
+	cbs := NewCallbackStore(ds)
+	// we give the final append store to the syncer
+	syncer := NewSyncer(l, cbs, crypto.chain, client)
 	c := &chainStore{
-		l:             l,
-		conf:          conf,
-		client:        client,
-		safe:          safe,
-		Store:         s,
-		done:          make(chan bool, 1),
-		ticker:        ticker,
-		newPartials:   make(chan partialInfo, defaultPartialChanBuffer),
-		newBeaconCh:   make(chan *chain.Beacon, defaultNewBeaconBuffer),
-		requestSync:   make(chan likeBeacon, defaultRequestSyncBuffer),
-		lastInserted:  make(chan *chain.Beacon, 1),
-		nonSyncBeacon: make(chan *chain.Beacon, 1),
+		l:                l,
+		conf:             conf,
+		client:           client,
+		CallbackStore:    cbs,
+		sync:             syncer,
+		crypto:           crypto,
+		done:             make(chan bool, 1),
+		ticker:           ticker,
+		newPartials:      make(chan partialInfo, defaultPartialChanBuffer),
+		catchedupBeacons: make(chan *chain.Beacon, 1),
```

Comment:
prefer `catchup` or `caughtup` to `catchedup`

Reply:
Lol sorry I lost my english yesterday